### PR TITLE
feat(classifier): enable LoRA auto-detection for intent classification

### LINKED ISSUE #724: Enable LoRA auto-detection for Intent/Category Classification
## Summary

Following the request from @Xunzhuo in [PR #718](https://github.com/vllm-project/semantic-router/pull/718#issuecomment-2493848361), this issue tracks the migration of Intent/Category Classification to support LoRA auto-detection.

## Background

Currently, PII detection already has LoRA auto-detection (implemented in PR #709), but Intent Classification does not. This creates an inconsistency where some classification features can leverage LoRA models while others cannot.

## Current Behavior (BEFORE)

**Problem**: Intent classification cannot automatically use LoRA models.

**How it works**:
1. Configuration has a `use_modernbert` flag that determines which initializer to use
2. System makes a hardcoded choice between two paths:
   - `use_modernbert: false` → Uses `LinearCategoryInitializer` (Traditional BERT only)
   - `use_modernbert: true` → Uses `ModernBertCategoryInitializer` (ModernBERT only)
3. Neither path can detect or use LoRA models automatically
4. Even if you point `model_id` to a LoRA model, it will fail because the initializer doesn't understand LoRA

**Example**:
```yaml
category_model:
  model_id: "models/lora_intent_classifier_bert-base-uncased_model"
  use_modernbert: false  # Will try Linear BERT, FAIL with LoRA model
```

## Expected Behavior (AFTER)

**Solution**: Intent classification should auto-detect LoRA models (just like PII does).

**How it should work**:
1. Single auto-detecting initializer that intelligently routes based on model type
2. Detection happens automatically by checking:
   - LoRA weights in model.safetensors file
   - Presence of lora_config.json
3. Smart fallback chain: LoRA → Traditional BERT → ModernBERT
4. The `use_modernbert` config flag becomes optional/ignored (backward compatible)
5. Zero configuration needed - just point to model path and system figures it out

**Example** (same config, now works):
```yaml
category_model:
  model_id: "models/lora_intent_classifier_bert-base-uncased_model"
  use_modernbert: false  # Ignored - auto-detection finds LoRA and uses it
```

## Implementation Notes

- Follow the same pattern as PII detection (PR #709)
- Update both Go layer (classifier.go) and Rust layer (init.rs, classify.rs)
- Add auto-detection test similar to PII
- Ensure backward compatibility with existing Traditional BERT and ModernBERT models
- LoRA intent models already exist in the models/ directory

## Related Work

- **PR #709**: PII LoRA auto-detection (reference implementation)
- **PR #718**: Fixed PII API confidence scores, comment requesting this migration
- **Issue #TBD**: Jailbreak Detection LoRA auto-detection (depends on this being merged first)

## Success Criteria

- LoRA intent models are automatically detected and used
- Traditional BERT and ModernBERT fallback paths still work
- No configuration changes required for users
- All existing tests continue to pass
- New tests demonstrate LoRA auto-detection working

### PR DESCRIPTION
**Pull request by yossiovadia**
## Summary
Enable LoRA auto-detection for intent/category classification, following the pattern from PII detection (PR #709).

Fixes #724

## Changes
- Updated `CategoryInitializer` to auto-detect LoRA models via `lora_config.json`
- Updated Rust FFI layer to support intent LoRA classification
- Removed hardcoded `useModernBERT` logic - now auto-detects model type
- Maintains backward compatibility with Traditional BERT and ModernBERT fallback
- Net reduction: 31 additions, 40 deletions

## Technical Details
**Go Layer (`classifier.go`):**
- Replaced separate initializers with unified `CategoryInitializerImpl`
- Auto-detection tries `InitCandleBertClassifier()` first (checks for `lora_config.json`)
- Falls back to ModernBERT if LoRA initialization fails

**Rust Layer (`init.rs`, `classify.rs`):**
- Added `LORA_INTENT_CLASSIFIER` static variable
- Updated `init_candle_bert_classifier` with intelligent model type detection
- Routes to `IntentLoRAClassifier::new()` for LoRA models
- Proper fallback chain: LoRA → Traditional BERT → ModernBERT

## Testing
- [x] Local testing with LoRA model (`lora_intent_classifier_bert-base-uncased_model`)
- [x] Auto-detection test passes (`TestIntentClassificationLoRAAutoDetection`)
- [x] E2E classification tests pass (03-classification-api-test.py)
- [x] Backward compatibility maintained
- [x] Pre-commit checks pass

## Configuration
No configuration changes needed - auto-detection works with existing E2E config:
\`\`\`yaml
classifier:
  category_model:
    model_id: "models/lora_intent_classifier_bert-base-uncased_model"
    use_cpu: true
\`\`\`

## Related
- Based on PII LoRA auto-detection pattern from PR #709
- Part of overall LoRA migration (Jailbreak to follow in separate PR for issue #725)
- Requested by @Xunzhuo in PR #718 comments

### CODE REVIEW

**Comment by Xunzhuo** [comment]
https://github.com/vllm-project/semantic-router/actions/runs/19620268281?pr=726 can you check the report, i saw the accuracy drops to 40%, any ideas?

**Comment by yossiovadia** [comment]
it seems that the LoRA intent classifier has 40.71% accuracy vs ModernBERT's ~~ 63.83% accuracy
hopefully we can retrain the Lora model, i'll investigate further tmrw

**Comment by Xunzhuo** [comment]
@yossiovadia thanks, so i think we should hold this PR for a while util LoRA reaches almost the same accuracy with full param

**Comment by yossiovadia** [comment]
I think that lora_config.json wasn't downloaded from HF, probably there's .downloaded file prevents re-downloading( if cache exists, it skips the download of the model )
the lora_config is in HF ( uploaded ~2 weeks ago ) 
however , testing it locally, though result improved, it's now 70% ( vs higher against modernbert. )
still looking ( but we need to clear the cache regardless. )

**Comment by yossiovadia** [comment]
yes, dont approve / merge yet, let me see if we can get at least same result as moderbert.

**Comment by yossiovadia** [comment]
Hey @Xunzhuo -
looking further ( if  lora_config.json exists ) the Lora is actually BETTER than the base ModernBERT (70% vs 64%)

If  u can clean the .download and re-kick this CI test again it would be great ? ( I'll only get back tmrw for this, it's getting late here )

**Comment by yossiovadia** [comment]
I'm back. testing the lora_config theory

**Comment by yossiovadia** [comment]
seems it still not fetching the config_lora, digging in.

**Comment by yossiovadia** [comment]
## Investigation Update: Accuracy Regression Confirmed

I've investigated the accuracy drop and can confirm there is a measurable regression when switching from ModernBERT to LoRA for intent classification.

### Findings

**Baseline (main branch with ModernBERT):**
- CI Run: https://github.com/vllm-project/semantic-router/actions/runs/19590181379
- Commit: c3ce62e (Nov 22, 2025)
- **Accuracy: 50.36%** (141/280 correct on MMLU-Pro dataset)

**Current PR (with LoRA):**
- CI Run: https://github.com/vllm-project/semantic-router/actions/runs/19620268281
- Commit: 4c9bc0c (Nov 24, 2025)
- **Accuracy: 40.71%** (114/280 correct on MMLU-Pro dataset)

### Analysis

1. **~10% accuracy drop** when switching from ModernBERT to LoRA (50% → 40%)
2. **Not caused by missing `lora_config.json`** - verified the file is downloaded correctly and LoRA auto-detection works as intended
3. **The LoRA model itself performs worse** on the MMLU-Pro academic question dataset used in E2E tests

### Next Steps

Holding this PR until we understand why the LoRA model underperforms ModernBERT on this dataset. Possible areas to investigate:
- Training data mismatch (LoRA training set vs. MMLU-Pro test set)
- Model capacity differences
- Hyperparameter tuning for LoRA
- Category mapping alignment

cc @Xunzhuo - confirming your observation about the accuracy drop. The auto-detection mechanism works correctly; the issue is with the LoRA model's performance on this specific dataset.

**Comment by yossiovadia** [comment]
Note to self ( and others ) 
useful command to fetch the accuracy from specific run -
```
gh run view 19590181379 --repo vllm-project/semantic-router --log 2>&1 | grep -E "Accuracy Rate|Domain classification test completed" | head -5
```

**Comment by yossiovadia** [comment]
Hey @Xunzhuo and @rootfs , 

PR #709 ( mine...)  enabled LoRA-based PII detection which has much higher sensitivity (90% success rate vs 27% with ModernBERT). While this fixed real PII detection issues (#647), it introduced a new problem: The LoRA PII model detects numbers in academic questions as PII:
"$1650" → B-ZIP_CODE (0.865 confidence)
"4111-1111-1111-1111" → B-US_SSN (0.867 confidence)
Various numbers → Phone numbers, credit cards, etc.


This seems to Impact on Domain Classification Tests:
LoRA intent classifier achieves ~80% accuracy (I have verified via direct testing)
But ~60% of domain test requests are blocked by PII false positives
Blocked requests don't return x-vsr-selected-category header
Test interprets missing headers as wrong classifications
Measured accuracy: 40.71% (should be ~80%)

few examples :
```
Business question: "A used car worth $1650 was purchased..." Detected: B-ZIP_CODE ('1650') with 0.865 confidence This is a price, not a ZIP code..
Credit card patterns in questions: "4111-1111-1111-1111" Detected as: B-US_SSN with 0.867 confidence 
Phone number patterns: Any sequence like "555-1234" Detected as: B-PHONE_NUMBER
```

im checking maybe a lower PII threshold now.

**Comment by Xunzhuo** [comment]
maybe we should separately test domain classification and pii/jailbreak

**Comment by yossiovadia** [comment]
Hey @Xunzhuo , I recommend have both. 
Keep this one for "real world scenario" ( what we have now)  and have independent for component test.

Note that  - 
The PII threshold change (0.7 → 0.9) demonstrates why we need both: ( last change I did in this PR)
```
Domain classification improved: 40.71% → 52.50% (+11.79%)
PII detection slightly decreased: 100% → 96% (-4%)
```

I'll see if I can find a better sweet-spot - higher classification (52.5%) and lower PII (96%) by playing with the threshold and let's keep that as our 'baseline' ?

Then i'll create independent as u sugested, WDYT?

**Comment by yossiovadia** [comment]
Small note - aibrix doesn't have its own values.yaml - it uses the default config. ( threshold is 0.7 ) that's why it'll show ~40%

**Comment by yossiovadia** [comment]
locally testing other values :
```
Threshold       | Domain Accuracy  | PII Accuracy
0.7             | 40.71% (114/280) | 100% (100/100) 
0.85            | 50.36% (141/280) | 99% (99/100) 
0.9             | 52.50% (147/280) | 96% (96/100)
```

For this test we should keep it 0.9.

@Xunzhuo , please see my comments above and let me know if you agree ( this PR is needed for enable LoRA auto-detection)

**Comment by yossiovadia** [comment]
last note on that topic, note that before this PR, using the modernbert, this was the results -
```
gh run view 19590181379 --repo vllm-project/semantic-router --log 2>&1 | grep -E "Accuracy Rate|Domain classification test completed" 
integration-test        UNKNOWN STEP    2025-11-22T04:36:26.9345363Z [Test] Domain classification test completed: 141/280 correct (50.36% accuracy)

```

File: src/semantic-router/pkg/classification/lora_auto_detection_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		t.Logf("✓ Classification works: category=%d, confidence=%.3f", result.Class, result.Confidence)
+		t.Logf("   Solution: Update CategoryInitializer to use InitCandleBertClassifier")
+	})
+}
+
+// TestPIIAlreadyHasAutoDetection shows PII detection already works with LoRA auto-detection
+func TestPIIAlreadyHasAutoDetection(t *testing.T) {
+	modelPath := "models/lora_pii_detector_bert-base-uncased_model"
```
Comment (rootfs) [inline src/semantic-router/pkg/classification/lora_auto_detection_test.go:92]: why this model path is different from the one above?

Comment (yossiovadia) [reply src/semantic-router/pkg/classification/lora_auto_detection_test.go:92]: Good catch! You're right - the PII test path is inconsistent. 

Line 92 should be: `../../../../models/lora_pii_detector_bert-base-uncased_model`

Currently it skips when the path isn't found (line 96), so it hasn't been failing - just silently skipping every run. The intent classification test (line 14) uses the correct relative path.

Since this is a test-only fix and doesn't affect the core functionality, I'll fix it in a follow-up PR to avoid invalidating your approval. Sound good?

Comment (rootfs) [reply src/semantic-router/pkg/classification/lora_auto_detection_test.go:92]: sgtm

**Comment by yossiovadia** [comment]
updated the new tests to use the LORA shows much better results - Domain classification went from 1% → 53% (50x improvement)  

@rootfs ,please review , we want this merged i assume.
